### ISSUE TITLE
Investigate possible use-cases for breakdown graphs according to the proposal

### ISSUE THREAD

**Issue by jahtalab** [elastic/apm-agent-rum-js#289]
The breakdown graphs proposal is in review status, so we should investigate if we can utilise it for RUM specific breakdowns

- [ ] Remove the timestamp from the payload sent to APM server
- [x] Provide detailed breakdown docs on each span type. elastic/apm-agent-rum-js#467 
- [x] Consider post transaction-end, creation of the breakdown metrics (#438   )
- [x] Revise the tests according to https://github.com/elastic/apm-agent-rum-js/issues/289#issuecomment-534999842 (elastic/apm-agent-rum-js#438 )
- [x] ~~Disable the breakdown metrics on transaction class by default (#434)~~ -> we use the config option

* https://docs.google.com/document/d/1-_LuC9zhmva0VvLgtI0KcHuLzNztPHbcM0ZdlcPUl64/edit?ts=5cd54da0#heading=h.j9rqr5k34x5v

**Comment by vigneshshanmugam** [comment]
Had a quick discussion with @jahtalab @roncohen @felixbarny, Here is the summary

+ Figure out what will be the valuable metric types that we can showcase to the user in the context of RUM.
+ For Example - We can identify the time spent on Network(TTFB, DNS etc) and Resources(Images, + CSS, JS) and add it to the breakdown
+ For non-sampled transactions, We can still calculate the aggregation either in agent side / APM server (Not decided yet)
+ The depth of the child in RUM is always 1, since we cannot differentiate between current frame and from other frames(iframe other domain).
+ For single page transactions, The types would just be presented in terms of resources -> we can pick individual types as JS, CSS and Images.

### Next steps
+ Go over some example transactions and figure out the self times based on the discussion as well as the algorithm presented on the proposal.

**Comment by vigneshshanmugam** [comment]
Had some discussion with @felixbarny  today. Here is how the breakdown would look for this page-load transaction 
<img width="1247" alt="Screenshot 2019-06-14 at 12 04 23" src="https://user-images.githubusercontent.com/3902525/59502005-0e27a680-8e9d-11e9-819c-e72f9a53ca83.png">

### Rough breakdown on different types

Key -  type of transaction
Value -  [count, self_time]  

Total duration - 4832 ms
```js
{
  app: [1, 1.869]
  network:  [3, 798],
  js: [5, 1118.8]
  css: [5, 242]
  fonts: [2, 495.5],
  img: [1, 308]
}
```

The above values would roughly be translated in to metric sets with span/transaction name and types 

```
{"metricset":{"transaction":{"name":"test","type":"page-load"},"samples":{"transaction.duration.count":{"value":1},"transaction.duration.sum":{"value":4832},"transaction.breakdown.count":{"value":1}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"navigation","subtype":"ttfb"},"samples":{"span.self_time.count":{"value":3},"span.self_time.sum":{"value":798}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"js"},"samples":{"span.self_time.count":{"value":5},"span.self_time.sum":{"value":1118.8}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"css"},"samples":{"span.self_time.count":{"value":5},"span.self_time.sum":{"value":242}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"fonts"},"samples":{"span.self_time.count":{"value":2},"span.self_time.sum":{"value":495.5}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"images"},"samples":{"span.self_time.count":{"value":1},"span.self_time.sum":{"value":308}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"app"},"samples":{"span.self_time.count":{"value":1},"span.self_time.sum":{"value":1.869}}}}

```

### Key differences

+ RUM agent groups the resource types(JS/CSS) and calculates the average time spent on downloading + executing the specific type on the browser. 
+ User timing spans https://github.com/elastic/apm-agent-rum-js/pull/276 will be added to the `app` type since its part of the application code. 
+ Excluded the `Parse document and exec sync scripts` span since it will hide inline scripts that are executed during page-load which should be considered as part of `app` self time. 
+ Right now we do not have much all the resource/network information for Route change/custom transactions. Keeping these fixed subtypes,  The solution would be generic enough here as well. 

### Open questions
+ Not really sure if we have to differentiate between DNS, SSL and Server Response time . In the example,I have grouped them and called it as `Network/TTFB`. 

Thoughts @roncohen

**Comment by roncohen** [comment]
1. I imagine having distinct types for DNS, Connect, WFFB (waiting for first byte) would be useful to have. 
1. I assume `app: [1, 1.869]` is actually `app: [1, 1869]` - there's no decimal point, right? if that the case this sums to just about the full page load time. However, as far as i understand the algorithm @felixbarny outlines, when you have two siblings with no children running at the same time, their self-time would both contribute to total sum. There's a lot of that going on here, but we still arrive at the full page load time. Could you elaborate what you did to arrive at these timings?

**Comment by vigneshshanmugam** [comment]
> I imagine having distinct types for DNS, Connect, WFFB (waiting for first byte) would be useful to have.

Yeah I had the same thought as well. But was wondering for custom/route-change transactions it might be harder to get those information. And also most of the page-load with cold .cache wont have those info. But yeah we could break it down. 

Yes you are right,  I have updated the data. So right now the way its calculated is

No of CSS Spans - 5
Sum of duration of all 5 CSS spans - 1210

@roncohen I totally screwed up the metric set values. I put the aggregated average value instead of the sum. The values of the JS, CSS and other resource types would look something like this

```
{"metricset":{"transaction":{"name":"test","type":"page-load"},"samples":{"transaction.duration.count":{"value":1},"transaction.duration.sum":{"value":4832},"transaction.breakdown.count":{"value":1}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"navigation","subtype":"ttfb"},"samples":{"span.self_time.count":{"value":3},"span.self_time.sum":{"value":798}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"js"},"samples":{"span.self_time.count":{"value":5},"span.self_time.sum":{"value":5594}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"css"},"samples":{"span.self_time.count":{"value":5},"span.self_time.sum":{"value":1210}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"fonts"},"samples":{"span.self_time.count":{"value":2},"span.self_time.sum":{"value":991}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"resource","subtype":"images"},"samples":{"span.self_time.count":{"value":1},"span.self_time.sum":{"value":308}}}}
{"metricset":{"transaction":{"name":"test","type":"page-load"},"span":{"type":"app"},"samples":{"span.self_time.count":{"value":1},"span.self_time.sum":{"value": 1869.7}}}}
```

**Transaction self time = total duration - average self time per group.**

**Comment by roncohen** [comment]
so the breakdown will say:

ttfb: 798ms
js: 5594ms 
css: 1210ms
fonts: 991ms
images: 308ms
app: 1869.7ms

total: 10,770.7ms 
ttfb: 798ms - 7.4%
js: 5594ms - 51.9%
css: 1210ms -  11.2%
fonts: 991ms - 9.2%
images: 308ms - 2.9%
app: 1869.7ms - 17.3%

ttfb should be 16% (798/4821), but it's actually only 7.4%. As the design doc explains, this is what happens when you have a bunch of things that happen at the same time like it does for RUM initial page loads. For backend (except Node.js), the model works because there isn't that much going on at the same time. For RUM initial page loads, it will be off most of the time.

Another issue is that the large gap after the last JS file has downloaded until "Parsing the document, executing sync. scripts" is finished, is probably spent parsing and evaluating JS files, but it gets attributed to "app" here. I think users would expect "js" to cover that. We can attribute it directly to css/js directly, but i think the aggregate would be way more useful if it reflected that this wasn't "app" but js/css.

It would be good to try and see if we can do something better.

**Comment by vigneshshanmugam** [comment]
> For RUM initial page loads, it will be off most of the time.

I agree with you totally on this. But for the async and RUM case, they don't naturally add up to 100% and We have to document explaining this in a better way. 

I am not sure if attributing to `js` might be the right choice here. Because there might be synchronous(inline) scripts from the application that might contribute to the parsing and script exec. There might be also other tasks that happen during that time like style recalculation & paint which is trigged due to those inline scripts. 

Also one more issue that we capture only the known resource types as of now https://github.com/elastic/apm-agent-rum-js/blob/0116f68aca3eccf5bc3fafcee6d13e87de8d656b/packages/rum-core/src/common/constants.js#L53. If the webpage has some other resources like video, we might be indirectly associating them to `js/css` which feels wrong. Instead `app` is a better choice.

**Comment by felixbarny** [comment]
> If the webpage has some other resources like video, we might be indirectly associating them to js/css which feels wrong. Instead app is a better choice.

Can't we set the type to `resource` and leave out the subtype?

> For backend (except Node.js), the model works because there isn't that much going on at the same time. For RUM initial page loads, it will be off most of the time.

This is a general issue with the concept. In Java, there are also frameworks which inherently work asynchronously/reactive. Even in mostly one thread per request models, users might take advantage of thread pools to make several external requests in parallel in order to combine the results or just take the fastest result.
I would like to avoid that the breakdown represents different things for different agents. So either we decide that the current design is fine and just needs some explanation or we come up with other ideas which can also be applied to other agents like node. One idea would be to break down only by the critical path. That would, for example, be the longest async child. However, that will yield a very unreliable result. For one transaction, downloading a CSS might be on the critical path, whereas for another one, it's an image download which took the longest.

Ofc, it's still worth to think about what the optimal solution for RUM would be. But we should then think about how to get as close to the optimal solution with a generic breakdown algorithm.

**Comment by roncohen** [comment]
@vigneshshanmugam if we could nest the <head> resources under `Parsing the documents, executing sync. scripts" (i know we don't directly have that information) and the applied the generic algorithm, the blank area would (more) correctly be attributed to "Parsing the documents, executing sync. scripts" which would make things slightly better.

**Comment by vigneshshanmugam** [comment]
@roncohen Yeah nice idea. Do you want the UI to reflect the relationship as well? I feel users might be confused since it does not align much with Devtools waterfall in the browser. We could simply assume the resource spans happening within the timeframe are child spans of the `Parsing the documents, executing sync. scripts` but we might need to rename it 

`Parsing the documents, executing sync. scripts` -> `Parsing the documents, executing scripts` 

since async scripts might also be dowloaded and executed within the time frame. 

Do you have any idea on what would be the type of the `Parsing the documents, executing sync. scripts` in the breakdown graph?

**Comment by vigneshshanmugam** [comment]
If we take the usual way of how breakdown would work for RUM agent and consider the same waterfall diagram, It would look something like this

```
Total duration - 4832
Network(DNS + SSL + HTTP) - 434 - 8.9%
Server Response Time / TTFB - 364 - 7.5%
Document Processing Time - 3553 - 73.53%
Page Render - 332 - 6.8 %
```

Again this applies only for page-load transactions. Another caveat is we wont be able to capture the time spent on the user application code unless they have instrumented the app using our API or [user timing api](https://developer.mozilla.org/en-US/docs/Web/API/User_Timing_API).

**Comment by jahtalab** [comment]
We've decided to moved this to next stack version.

**Comment by roncohen** [comment]
Yes i agree. We need to do better than applying the generic algorithm here. However, the details about _how_ we derive things like `Document processing time` and `page render` definitely do not seem straight forward. @vigneshshanmugam will you and @jahtalab look into which meaning full steps we can slice the page load into?

**Comment by vigneshshanmugam** [comment]
All of those timings mentioned in the breakdown are part of the navigation timing API https://developer.mozilla.org/en-US/docs/Web/API/Navigation_timing_API and its easy to gather and group them to our needs.

**Comment by roncohen** [comment]
Ah, that's great. Let's move forward with those 👍

**Comment by jahtalab** [comment]
Just had a call with @vigneshshanmugam and we decided the following:
* We should add some tests for the selfTime calculation in the `transaction.spec.js`
* We decided to remove the performance.now() patch in the `breakdown.spec` and instead set the constant values for selfTime plus add the `start` and `end` to the options parameter.
* We talked about the post transaction end creation of the breakdown and discussed the following algorithm, we need to test this agains our current test cases:

```
function calculateSelfTime({ spans, _start, _end }) {
    var  lastContinuousEnd
    var selfTime = 0
    for (let i = 0; i < spans.length; i++) {
        const span = spans[i];
        if (!lastContinuousStart || !lastContinuousEnd) {
            lastContinuousEnd = span._end
            if (span._start > _start) {
                selfTime += span._start - _start
            }
            continue
        }

        if (span._start > lastContinuousEnd) {
            selfTime += span._start - lastContinuousEnd
            lastContinuousEnd = span._end
            continue
        } else if (span._end > lastContinuousEnd) {
            lastContinuousEnd = span._end
        }
    }

    if (lastContinuousEnd > _end) {
        selfTime += lastContinuousEnd - _end
    }

    return selfTime
}
```
